libs/types-rs/src/cacvote/mod.rs

```diff
@@ -46,9 +46,72 @@ pub struct SignedObject {
 }
 
 impl SignedObject {
+    #[cfg(feature = "openssl")]
+    pub fn from_payload(
+        payload: &Payload,
+        certificates: Vec<openssl::x509::X509>,
+        private_key: &openssl::pkey::PKeyRef<openssl::pkey::Private>,
+    ) -> color_eyre::Result<Self> {
+        let mut signer =
+            openssl::sign::Signer::new(openssl::hash::MessageDigest::sha256(), private_key)?;
+        let payload = serde_json::to_vec(payload)?;
+        signer.update(&payload)?;
+        let signature = signer.sign_to_vec()?;
+
+        let certificates = certificates
+            .iter()
+            .map(|cert| cert.to_pem())
+            .collect::<Result<Vec<_>, _>>()?
+            .concat();
+
+        Ok(Self {
+            payload,
+            certificates,
+            signature,
+        })
+    }
+
     pub fn try_to_inner(&self) -> Result<Payload, serde_json::Error> {
         serde_json::from_slice(&self.payload)
     }
+
+    #[cfg(feature = "openssl")]
+    pub fn to_x509(&self) -> Result<Vec<openssl::x509::X509>, openssl::error::ErrorStack> {
+        openssl::x509::X509::stack_from_pem(&self.certificates)
```

Comment:
FYI @benadida, it looks like OpenSSL has built-in support for getting a "stack" of X509 certs from a buffer, which aligns well with our discussion earlier. I'm going to assume that's what we're dealing with when referencing the `certificates` field.

Reply:
yes that's exactly it!